src/instrument.cpp

```diff
@@ -48,77 +49,109 @@ class Constant {
+            case JVM_CONSTANT_Integer:
+            case JVM_CONSTANT_Float:
+            case JVM_CONSTANT_Fieldref:
+            case JVM_CONSTANT_Methodref:
+            case JVM_CONSTANT_InterfaceMethodref:
+            case JVM_CONSTANT_NameAndType:
+            case JVM_CONSTANT_Dynamic:
+            case JVM_CONSTANT_InvokeDynamic:
                 return 4;
-            case CONSTANT_Long:
-            case CONSTANT_Double:
+            case JVM_CONSTANT_Long:
+            case JVM_CONSTANT_Double:
                 return 8;
-            case CONSTANT_Class:
-            case CONSTANT_String:
-            case CONSTANT_MethodType:
+            case JVM_CONSTANT_Class:
+            case JVM_CONSTANT_String:
+            case JVM_CONSTANT_MethodType:
             case CONSTANT_Module:
             case CONSTANT_Package:
                 return 2;
-            case CONSTANT_MethodHandle:
+            case JVM_CONSTANT_MethodHandle:
                 return 3;
             default:
                 return 0;
         }
     }
 
     bool equals(const char* value, u16 len) {
-        return _tag == CONSTANT_Utf8 && info() == len && memcmp(_info + 2, value, len) == 0;
+        return _tag == JVM_CONSTANT_Utf8 && info() == len && memcmp(utf8(), value, len) == 0;
     }
 
     bool matches(const char* value, u16 len) {
         if (len > 0 && value[len - 1] == '*') {
-            return _tag == CONSTANT_Utf8 && info() >= len - 1 && memcmp(_info + 2, value, len - 1) == 0;
+            return _tag == JVM_CONSTANT_Utf8 && info() >= len - 1 && memcmp(utf8(), value, len - 1) == 0;
         }
         return equals(value, len);
     }
+
+    static u8 parameterSlots(const char* method_sig) {
+        u8 count = 0;
+        for (const char* c = method_sig + 1; *c != ')'; ++c) {
+            if (*c == 'L') {
+                ++count;
+                while (*(++c) != ';');
+            } else if (*c == '[') {
+                ++count;
+                while (*(++c) == '[');
+                if (*c == 'L') while (*(++c) != ';');
+            } else if (*c == 'J' || *c == 'D') {
+                count += 2;
+            } else {
+                ++count;
+            }
+        }
+        return count;
+    }
 };
 
 enum Scope {
     SCOPE_CLASS,
     SCOPE_FIELD,
     SCOPE_METHOD,
-    SCOPE_REWRITE_METHOD,
-    SCOPE_REWRITE_CODE
+    SCOPE_REWRITE_METHOD
 };
 
 enum PatchConstants {
-    EXTRA_CONSTANTS = 6,
-    EXTRA_BYTECODES = 4,
-    EXTRA_STACKMAPS = 1
+    EXTRA_CONSTANTS = 16,
+    // Entry which does not track start time
+    EXTRA_BYTECODES_SIMPLE_ENTRY = 4,
+    // System.nanoTime and lstore
+    EXTRA_BYTECODES_ENTRY = 8,
+    // lload and recordExit(startTime)
+    EXTRA_BYTECODES_EXIT = 8,
+    // *load_i or *store_i to *load/*store
+    EXTRA_BYTECODES_INDEXED = 4
 };
 
-
 class BytecodeRewriter {
   private:
     const u8* _src;
     const u8* _src_limit;
 
+    const char* _class_name;
+    u16 _class_name_len;
```

Comment:
Maybe `Constant*`? You'll have to deal everywhere with one variable instead of two.

Reply:
69cc6995b5aeed4a4569a039d87557f9d5c44602